src/controller/java/AndroidCommissioningWindowOpener.cpp

```diff
@@ -0,0 +1,176 @@
+    auto * opener = new AndroidCommissioningWindowOpener(controller, jcallback);
+    if (opener == nullptr)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    CHIP_ERROR err = opener->CommissioningWindowOpener::OpenBasicCommissioningWindow(
+        deviceId, timeout, &opener->mOnOpenBasicCommissioningWindowCallback);
+    if (err != CHIP_NO_ERROR)
+    {
+        delete opener;
+    }
+    // Else will clean up when the callback is called.
+    return err;
+}
+
+CHIP_ERROR AndroidCommissioningWindowOpener::OpenCommissioningWindow(DeviceController * controller, NodeId deviceId,
+                                                                     Seconds16 timeout, uint32_t iteration, uint16_t discriminator,
+                                                                     Optional<uint32_t> setupPIN, Optional<ByteSpan> salt,
+                                                                     jobject jcallback, SetupPayload & payload,
+                                                                     bool readVIDPIDAttributes)
+{
+    // Not using Platform::New because we want to keep our constructor private.
+    auto * opener = new AndroidCommissioningWindowOpener(controller, jcallback);
+    if (opener == nullptr)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    CHIP_ERROR err = opener->CommissioningWindowOpener::OpenCommissioningWindow(
+        deviceId, timeout, iteration, discriminator, setupPIN, salt, &opener->mOnOpenCommissioningWindowCallback, payload,
+        readVIDPIDAttributes);
+    if (err != CHIP_NO_ERROR)
+    {
+        delete opener;
+    }
+    // Else will clean up when the callback is called.
+    return err;
+}
+
+void AndroidCommissioningWindowOpener::OnOpenCommissioningWindowResponse(void * context, NodeId deviceId, CHIP_ERROR status,
+                                                                         chip::SetupPayload payload)
+{
+    auto * self    = static_cast<AndroidCommissioningWindowOpener *>(context);
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (status == CHIP_NO_ERROR)
+    {
+        std::string QRCode;
+        std::string manualPairingCode;
+
+        err = ManualSetupPayloadGenerator(payload).payloadDecimalStringRepresentation(manualPairingCode);
+        err = QRCodeSetupPayloadGenerator(payload).payloadBase38Representation(QRCode);
+        if (self->mJavaCallback != nullptr && sOnSuccessMethod != nullptr)
+        {
+            JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
+            UtfString jManualPairingCode(env, manualPairingCode.c_str());
+            UtfString jQRCode(env, QRCode.c_str());
+            env->CallVoidMethod(self->mJavaCallback, sOnSuccessMethod, static_cast<jlong>(deviceId), jManualPairingCode.jniValue(),
+                                jQRCode.jniValue());
+        }
+    }
+    else
+    {
+        if (self->mJavaCallback != nullptr && sOnErrorMethod != nullptr)
```

Comment:
this if condition is nested on both branches. Could we move it outside as a precondition and do early check on it?

Reply:
mJavaCallback variable check will move to second line in this function.